### PR TITLE
SDFV: Draggable graph elements

### PR DESCRIPTION
**Pull request by phschaad**
- Nodes and edges can both be dragged
- Moving/panning of the view was moved from primary mouse button to secondary mouse button (for now?)
- Movement of nodes is restricted to remain within their parent element, e.g. state or nested SDFG

### CODE REVIEW

File: diode/webclient/renderer.js [hunk 1]
```diff
@@ -205,6 +205,161 @@ class CanvasManager {
...
+
+            // Make sure we don't move our element outside its parent's
+            // bounding box
+            const target_x = el.x + dx;
+            const target_y = el.y + dy;
+            if (target_x <= min_x || target_x >= max_x) dx = 0;
+            if (target_y <= min_y || target_y >= max_y) dy = 0;
+        }
+
+        if (el.data.type && (el.data.type === 'Memlet' || el.data.type === 'InterstateEdge')) {
```
Comment (tbennun) [inline diode/webclient/renderer.js:276]: you can check if it's a `typeof Edge` as far as I remember

File: diode/webclient/renderer.js [hunk 2]
```diff
@@ -205,6 +205,161 @@ class CanvasManager {
...
+            if (el.points[2]) {
+                // Only allow dragging, if the memlet is 'making a curve'
+                el.points[1].x += dx;
+                el.points[1].y += dy;
+            }
+            // Don't do any of the other stuff, it doesn't apply here
```
Comment (tbennun) [inline diode/webclient/renderer.js:282]:
```suggestion
            // The rest of the method doesn't apply here
```

File: diode/webclient/renderer.js [hunk 3]
```diff
@@ -205,6 +205,161 @@ class CanvasManager {
...
+                node.out_connectors.forEach(c => {
+                    c.x += dx;
+                    c.y += dy;
+                });
+        }
+
+        // Allow recursive translation of nested SDFGs
+        function translate_recursive(ng) {
+            ng.nodes().forEach(state_id => {
+                const state = ng.node(state_id);
```
Comment (tbennun) [inline diode/webclient/renderer.js:307]: why not use `let`?

File: diode/webclient/renderer.js [hunk 4]
```diff
@@ -1183,7 +1412,7 @@ class SDFGRenderer {
                 return false;
             } else {
                 this.drag_start = null;
-                if (event.buttons & 1)
+                if (event.buttons & 1 || event.buttons & 2)
```
Comment (tbennun) [inline diode/webclient/renderer.js:1415]: are you sure it should be 2 rather than 4 here?

File: diode/webclient/renderer.js (hunk 3 above)
Comment (phschaad) [reply diode/webclient/renderer.js:307]: I assumed `prefer-const`, is let generally preferred?

Comment (tbennun) [reply diode/webclient/renderer.js:307]: you can use const if it is correct. I normally don't.

File: diode/webclient/renderer.js (hunk 1 above)
Comment (phschaad) [reply diode/webclient/renderer.js]: assuming you meant `instanceof`, yes that works, thank you!

**Comment by phschaad** [comment]
Thank you!

**Pull request merged**